File: src/FarmEngine.Schemas/Primitives.cs
namespace FarmEngine.Schemas;

// Port of packages/engine-schemas/src/primitives.ts.
//
// Primitive enums shared by content, project and save schemas. zod string
// enums stay plain `string`s in C# (content is open-ended and JSON must
// round-trip); each enum gets a static class of constants plus `All`.

/// <summary>TS <c>TileTypeSchema</c>.</summary>
public static class TileTypes
{
    public const string Grass = "grass";
    public const string Soil = "soil";
    public const string Water = "water";
    public const string Path = "path";
    public const string Wall = "wall";
    public const string Door = "door";
    public const string Floor = "floor";

    public static readonly IReadOnlyList<string> All = [Grass, Soil, Water, Path, Wall, Door, Floor];
}

/// <summary>TS <c>DirectionSchema</c>.</summary>
public static class Directions
{
    public const string Up = "up";
    public const string Down = "down";
    public const string Left = "left";
    public const string Right = "right";

    public static readonly IReadOnlyList<string> All = [Up, Down, Left, Right];
}

/// <summary>TS <c>ItemTypeSchema</c>.</summary>
public static class ItemTypes
{
    public const string Seed = "seed";
    public const string Tool = "tool";
    public const string Crop = "crop";
    public const string Gift = "gift";
    public const string Quest = "quest";
    public const string Fertilizer = "fertilizer";
    public const string Material = "material";
    public const string Fish = "fish";

    public static readonly IReadOnlyList<string> All = [Seed, Tool, Crop, Gift, Quest, Fertilizer, Material, Fish];
}

/// <summary>TS <c>ToolTypeSchema</c>.</summary>
public static class ToolTypes
{
    public const string WateringCan = "watering-can";
    public const string Hoe = "hoe";
    public const string Axe = "axe";
    public const string Pickaxe = "pickaxe";
    public const string Scythe = "scythe";
    public const string FishingRod = "fishing-rod";

    public static readonly IReadOnlyList<string> All = [WateringCan, Hoe, Axe, Pickaxe, Scythe, FishingRod];
}

/// <summary>TS <c>EditorModeSchema</c>.</summary>
public static class EditorModes
{
    public const string Tiles = "tiles";
    public const string Npcs = "npcs";
    public const string Items = "items";
    public const string Events = "events";
    public const string Play = "play";
    public const string Quests = "quests";

    public static readonly IReadOnlyList<string> All = [Tiles, Npcs, Items, Events, Play, Quests];
}

/// <summary>TS <c>QuestStatusSchema</c>.</summary>
public static class QuestStatuses
{
    public const string NotStarted = "not-started";
    public const string Active = "active";
    public const string Completed = "completed";
    public const string Failed = "failed";

    public static readonly IReadOnlyList<string> All = [NotStarted, Active, Completed, Failed];
}

/// <summary>TS <c>QuestObjectiveTypeSchema</c>.</summary>
public static class QuestObjectiveTypes
{
    public const string Collect = "collect";
    public const string Harvest = "harvest";
    public const string Talk = "talk";
    public const string Visit = "visit";
    public const string Craft = "craft";
    public const string Gift = "gift";

    public static readonly IReadOnlyList<string> All = [Collect, Harvest, Talk, Visit, Craft, Gift];
}

/// <summary>TS <c>SoilStateSchema</c>.</summary>
public static class SoilStates
{
    public const string Dry = "dry";
    public const string Watered = "watered";
    public const string Fertilized = "fertilized";
    public const string Tilled = "tilled";

    public static readonly IReadOnlyList<string> All = [Dry, Watered, Fertilized, Tilled];
}

/// <summary>TS <c>CropQualitySchema</c>.</summary>
public static class CropQualities
{
    public const string Normal = "normal";
    public const string Silver = "silver";
    public const string Gold = "gold";
    public const string Iridium = "iridium";

    public static readonly IReadOnlyList<string> All = [Normal, Silver, Gold, Iridium];
}

/// <summary>
/// TS <c>CropMutationSchema</c>: one of these or <c>null</c> (so C# fields
/// holding it are <c>string?</c> written as <c>null</c>).
/// </summary>
public static class CropMutations
{
    public const string Giant = "giant";
    public const string Golden = "golden";
    public const string Ancient = "ancient";

    public static readonly IReadOnlyList<string> All = [Giant, Golden, Ancient];
}

public static class PrimitivesSchema
{
    /// <summary>
    /// Seasons are open strings (creator-defined calendars, M9) so custom season
    /// ids validate the same way custom crops do. <c>CLASSIC_SEASONS</c> enumerates
    /// the four seasons the built-in calendar ships with, for defaults/editors.
    /// </summary>
    public static readonly IReadOnlyList<string> ClassicSeasons = ["spring", "summer", "fall", "winter"];

    /// <summary>
    /// Crop identifiers are open strings so that custom/modded crops share the
    /// same pipeline as built-in ones. <c>BUILTIN_CROP_IDS</c> enumerates the crops
    /// shipped with the default content pack.
    /// </summary>
    public static readonly IReadOnlyList<string> BuiltinCropIds =
    [
        "wheat", "corn", "tomato", "carrot", "potato",
        "strawberry", "pumpkin", "cauliflower", "blueberry",
    ];
}

File: crates/farm-sim/src/schema/primitives.rs
//! Port of `Primitives.cs` (packages/engine-schemas/src/primitives.ts).
//!
//! Primitive enums shared by content, project and save schemas. zod string enums stay plain
//! `String`s (content is open-ended and JSON must round-trip); each enum gets a module of
//! constants plus `ALL`.

/// TS `TileTypeSchema`.
pub mod tile_types {
    pub const GRASS: &str = "grass";
    pub const SOIL: &str = "soil";
    pub const WATER: &str = "water";
    pub const PATH: &str = "path";
    pub const WALL: &str = "wall";
    pub const DOOR: &str = "door";
    pub const FLOOR: &str = "floor";

    pub const ALL: &[&str] = &[GRASS, SOIL, WATER, PATH, WALL, DOOR, FLOOR];
}

/// TS `DirectionSchema`.
pub mod directions {
    pub const UP: &str = "up";
    pub const DOWN: &str = "down";
    pub const LEFT: &str = "left";
    pub const RIGHT: &str = "right";

    pub const ALL: &[&str] = &[UP, DOWN, LEFT, RIGHT];
}

/// TS `ItemTypeSchema`.
pub mod item_types {
    pub const SEED: &str = "seed";
    pub const TOOL: &str = "tool";
    pub const CROP: &str = "crop";
    pub const GIFT: &str = "gift";
    pub const QUEST: &str = "quest";
    pub const FERTILIZER: &str = "fertilizer";
    pub const MATERIAL: &str = "material";
    pub const FISH: &str = "fish";

    pub const ALL: &[&str] = &[SEED, TOOL, CROP, GIFT, QUEST, FERTILIZER, MATERIAL, FISH];
}

/// TS `ToolTypeSchema`.
pub mod tool_types {
    pub const WATERING_CAN: &str = "watering-can";
    pub const HOE: &str = "hoe";
    pub const AXE: &str = "axe";
    pub const PICKAXE: &str = "pickaxe";
    pub const SCYTHE: &str = "scythe";
    pub const FISHING_ROD: &str = "fishing-rod";

    pub const ALL: &[&str] = &[WATERING_CAN, HOE, AXE, PICKAXE, SCYTHE, FISHING_ROD];
}

/// TS `EditorModeSchema`.
pub mod editor_modes {
    pub const TILES: &str = "tiles";
    pub const NPCS: &str = "npcs";
    pub const ITEMS: &str = "items";
    pub const EVENTS: &str = "events";
    pub const PLAY: &str = "play";
    pub const QUESTS: &str = "quests";

    pub const ALL: &[&str] = &[TILES, NPCS, ITEMS, EVENTS, PLAY, QUESTS];
}

/// TS `QuestStatusSchema`.
pub mod quest_statuses {
    pub const NOT_STARTED: &str = "not-started";
    pub const ACTIVE: &str = "active";
    pub const COMPLETED: &str = "completed";
    pub const FAILED: &str = "failed";

    pub const ALL: &[&str] = &[NOT_STARTED, ACTIVE, COMPLETED, FAILED];
}

/// TS `QuestObjectiveTypeSchema`.
pub mod quest_objective_types {
    pub const COLLECT: &str = "collect";
    pub const HARVEST: &str = "harvest";
    pub const TALK: &str = "talk";
    pub const VISIT: &str = "visit";
    pub const CRAFT: &str = "craft";
    pub const GIFT: &str = "gift";

    pub const ALL: &[&str] = &[COLLECT, HARVEST, TALK, VISIT, CRAFT, GIFT];
}

/// TS `SoilStateSchema`.
pub mod soil_states {
    pub const DRY: &str = "dry";
    pub const WATERED: &str = "watered";
    pub const FERTILIZED: &str = "fertilized";
    pub const TILLED: &str = "tilled";

    pub const ALL: &[&str] = &[DRY, WATERED, FERTILIZED, TILLED];
}

/// TS `CropQualitySchema`.
pub mod crop_qualities {
    pub const NORMAL: &str = "normal";
    pub const SILVER: &str = "silver";
    pub const GOLD: &str = "gold";
    pub const IRIDIUM: &str = "iridium";

    pub const ALL: &[&str] = &[NORMAL, SILVER, GOLD, IRIDIUM];
}

/// TS `CropMutationSchema`: one of these or `null` (so fields holding it are `Option<String>`
/// written as `null`).
pub mod crop_mutations {
    pub const GIANT: &str = "giant";
    pub const GOLDEN: &str = "golden";
    pub const ANCIENT: &str = "ancient";

    pub const ALL: &[&str] = &[GIANT, GOLDEN, ANCIENT];
}

/// Seasons are open strings (creator-defined calendars, M9) so custom season ids validate the
/// same way custom crops do. `CLASSIC_SEASONS` enumerates the four seasons the built-in calendar
/// ships with, for defaults/editors.
pub const CLASSIC_SEASONS: &[&str] = &["spring", "summer", "fall", "winter"];

/// Crop identifiers are open strings so that custom/modded crops share the same pipeline as
/// built-in ones. `BUILTIN_CROP_IDS` enumerates the crops shipped with the default content pack.
pub const BUILTIN_CROP_IDS: &[&str] =
    &["wheat", "corn", "tomato", "carrot", "potato", "strawberry", "pumpkin", "cauliflower", "blueberry"];
